fedimint-client/src/sm/executor.rs

```diff
@@ -390,181 +410,235 @@ impl ExecutorInner {
+                                                    let _ = sm_update_tx.send(state);
+                                                });
+                                                Ok(ActiveOrInactiveState::Active {
+                                                    dyn_state: new_state,
+                                                    meta: v,
+                                                })
+                                            }
+                                        })
+                                    },
+                                    None,
                                 )
-                                .await;
+                                .await
+                                .expect("autocommit should keep trying to commit (max_attempt: None) and body doesn't return errors");
 
-                                let context = &self
-                                    .module_contexts
-                                    .get(&state.module_instance_id())
-                                    .expect("Unknown module");
-
-                                let global_context = global_context_gen(
-                                    state.module_instance_id(),
-                                    state.operation_id(),
-                                );
-                                if new_state.is_terminal(context, &global_context) {
-                                    // TODO: log state machine id or something
-                                    debug!("State machine reached terminal state");
-                                    let k = InactiveStateKey::from_state(new_state.clone());
-                                    let v = ActiveState::new().into_inactive();
-                                    dbtx.insert_entry(&k, &v).await;
-                                    Ok(ActiveOrInactiveState::Inactive {
-                                        dyn_state: new_state,
-                                    })
-                                } else {
-                                    let k = ActiveStateKey::from_state(new_state.clone());
-                                    let v = ActiveState::new();
-                                    dbtx.insert_entry(&k, &v).await;
-                                    Ok(ActiveOrInactiveState::Active {
-                                        dyn_state: new_state,
-                                        active_state: v,
-                                    })
+
+                            match &outcome {
+                                ActiveOrInactiveState::Active {
+                                    dyn_state,
+                                    meta: _,
+                                } => {
+                                    sm_update_tx.send(dyn_state.clone()).expect("can't fail: we are the receiving end");
+                                    notifier.notify(dyn_state.clone());
```

Comment:
It's a bit annoying that we now have two notification systems: one for active states only and one for all state transitions. Ideally these could be combined eventually imo.

Reply:
Not a big issue, and not clear if it merging them will be better/possible. If there are N notifications needed then there we send it N times. At least the executor channel is mpsc.